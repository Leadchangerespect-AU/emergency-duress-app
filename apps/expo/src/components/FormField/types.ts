export interface FormFieldProps {
  label: string;
  isRequired?: boolean;
  isDisabled?: boolean;
  errors: (undefined | false | null | string)[];
}
