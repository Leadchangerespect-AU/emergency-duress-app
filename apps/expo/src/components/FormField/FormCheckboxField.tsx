import {
  FormControl,
  FormControlError,
  FormControlErrorIcon,
  FormControlErrorText,
  FormControlLabel,
} from "@/components/FormControl";
import Checkbox, { CheckboxProps } from "../Checkbox";
import { FormFieldProps } from "./types";

export const FormCheckboxField = ({
  isRequired,
  errors,
  isDisabled,
  children,
  ...props
}: Omit<FormFieldProps, "label"> & CheckboxProps) => {
  const hasErrors = errors.length > 0;

  return (
    <FormControl
      isInvalid={hasErrors}
      isRequired={isRequired}
      isDisabled={isDisabled}
    >
      <Checkbox {...props}>{children}</Checkbox>
      {hasErrors && (
        <FormControlError>
          <FormControlErrorIcon />
          <FormControlErrorText>{errors.join(", ")}</FormControlErrorText>
        </FormControlError>
      )}
    </FormControl>
  );
};
