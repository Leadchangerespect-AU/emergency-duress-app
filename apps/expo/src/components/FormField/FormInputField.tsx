import { useId } from "react";
import { Platform, TextInputProps } from "react-native";

import {
  FormControl,
  FormControlError,
  FormControlErrorIcon,
  FormControlErrorText,
  FormControlLabel,
} from "@/components/FormControl";
import Input from "@/components/Input";
import { FormFieldProps } from "./types";

export const FormInputField = ({
  label,
  isRequired,
  isDisabled,
  errors,
  ...props
}: FormFieldProps & TextInputProps) => {
  const hasErrors = errors.length > 0;

  const labelId = useId();
  const inputId = useId();
  return (
    <FormControl
      isInvalid={hasErrors}
      isRequired={isRequired}
      isDisabled={isDisabled}
    >
      <FormControlLabel
        id={labelId}
        {...Platform.select({
          web: {
            htmlFor: inputId,
          },
        })}
      >
        {label}
      </FormControlLabel>
      <Input
        {...props}
        aria-labelledby={labelId}
        accessibilityLabel={label}
        id={inputId}
      />
      {hasErrors && (
        <FormControlError>
          <FormControlErrorIcon />
          <FormControlErrorText>{errors.join(", ")}</FormControlErrorText>
        </FormControlError>
      )}
    </FormControl>
  );
};
