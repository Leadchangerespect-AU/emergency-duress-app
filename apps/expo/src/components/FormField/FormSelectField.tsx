import { useId } from "react";
import { Platform } from "react-native";

import {
  FormControl,
  FormControlError,
  FormControlErrorIcon,
  FormControlErrorText,
  FormControlLabel,
} from "@/components/FormControl";
import Select from "@/components/Select";
import { SelectProps } from "@/components/Select/types";
import { FormFieldProps } from "./types";

export const FormSelectField = ({
  label,
  isRequired,
  isDisabled,
  errors,
  ...props
}: FormFieldProps & SelectProps) => {
  const hasErrors = errors.length > 0;

  const labelId = useId();
  const selectId = useId();
  return (
    <FormControl
      isInvalid={hasErrors}
      isRequired={isRequired}
      isDisabled={isDisabled}
    >
      <FormControlLabel
        id={labelId}
        {...Platform.select({
          web: {
            htmlFor: selectId,
          },
        })}
      >
        {label}
      </FormControlLabel>
      <Select
        {...props}
        aria-labelledby={labelId}
        accessibilityLabel={label}
        id={selectId}
      />
      {hasErrors && (
        <FormControlError>
          <FormControlErrorIcon />
          <FormControlErrorText>{errors.join(", ")}</FormControlErrorText>
        </FormControlError>
      )}
    </FormControl>
  );
};
