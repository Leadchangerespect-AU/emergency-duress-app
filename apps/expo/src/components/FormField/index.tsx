import { FormInputField } from "./FormInputField";
import { FormSelectField } from "./FormSelectField";

export { FormInputField, FormSelectField };
