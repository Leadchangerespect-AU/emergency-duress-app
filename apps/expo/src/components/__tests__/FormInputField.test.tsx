import React from "react";
import { fireEvent, render, screen } from "@testing-library/react-native";

import { FormInputField } from "../FormField/FormInputField";

// Mock the Ionicons import used by FormControlErrorIcon
jest.mock("@expo/vector-icons", () => ({
  Ionicons: "",
}));

describe("FormInputField", () => {
  it("renders with required label", () => {
    render(
      <FormInputField
        label="Test Input"
        isRequired
        errors={[]}
        placeholder="Enter text"
      />,
    );

    const input = screen.getByLabelText("Test Input");
    expect(input).toBeTruthy();
    expect(input.props["aria-required"]).toBe(true);
  });

  it("displays error messages", () => {
    const errors = ["Required field", "Invalid format"];
    render(<FormInputField label="Test Input" errors={errors} />);

    // Find the combined error message
    expect(screen.getByText("Required field, Invalid format")).toBeTruthy();
  });

  it("does not set aria-invalid when there are no errors", () => {
    render(<FormInputField label="Test Input" errors={[]} />);

    const input = screen.getByLabelText("Test Input");
    expect(input.props["aria-invalid"]).toBeFalsy();
  });

  it("sets aria-invalid when there are errors", () => {
    render(<FormInputField label="Test Input" errors={["Error!"]} />);

    const input = screen.getByLabelText("Test Input");
    expect(input.props["aria-invalid"]).toBe(true);
  });

  it("handles text input changes", () => {
    const onChangeText = jest.fn();
    render(
      <FormInputField
        label="Test Input"
        errors={[]}
        onChangeText={onChangeText}
        placeholder="Enter text"
      />,
    );

    const input = screen.getByPlaceholderText("Enter text");
    fireEvent.changeText(input, "test value");
    expect(onChangeText).toHaveBeenCalledWith("test value");
  });

  it("applies disabled state", () => {
    render(<FormInputField label="Test Input" errors={[]} editable={false} />);

    const input = screen.getByLabelText("Test Input");
    expect(input.props.editable).toBe(false);
  });

  it("matches snapshot", () => {
    render(<FormInputField label="Test Input" errors={[]} />);
    expect(screen.toJSON).toMatchSnapshot();
  });

  it("matches snapshot with error state", () => {
    render(
      <FormInputField
        label="Test Input"
        isRequired
        errors={["Required field"]}
        placeholder="Enter text"
      />,
    );
    expect(screen.toJSON()).toMatchSnapshot();
  });
});
