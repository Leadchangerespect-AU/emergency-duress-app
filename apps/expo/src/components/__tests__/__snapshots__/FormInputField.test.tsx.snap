// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormInputField matches snapshot 1`] = `[Function]`;

exports[`FormInputField matches snapshot with error state 1`] = `
<View
  className="my-2"
>
  <Text
    className="text-base text-neutral-800 font-semibold"
    id=":re:"
  >
    Test Input
    <Text
      className="ml-0.5"
    >
      *
    </Text>
  </Text>
  <TextInput
    accessibilityLabel="Test Input"
    aria-invalid={true}
    aria-labelledby=":re:"
    aria-required={true}
    className="block w-full rounded-md bg-white px-3 py-2.5 text-base text-gray-900 outline outline-1 -outline-offset-1 outline-gray-300 placeholder:text-gray-400 focus:outline focus:outline-2 focus:-outline-offset-2 focus:outline-accent sm:text-sm/6 border-red-500"
    editable={true}
    id=":rf:"
    placeholder="Enter text"
    selectTextOnFocus={true}
  />
  <View
    className="flex flex-row items-center mt-2 gap-2"
  >
    <
      color="#ef4444"
      name="alert-circle-outline"
      size={24}
    />
    <Text
      className="text-red-500 text-sm"
    >
      Required field
    </Text>
  </View>
</View>
`;
